### PR TITLE
fix: leaking route handlers when times is used in Page.route

### LINKED ISSUE #9215: [BUG] Clear finished Page.route({times}) handlers once they are done

### PR DESCRIPTION
**Pull request by mxschmitt**
Fixes https://github.com/microsoft/playwright/issues/9215

### CODE REVIEW

File: src/client/page.ts [hunk 1]
```diff
@@ -169,13 +169,22 @@ export class Page extends ChannelOwner<channels.PageChannel, channels.PageInitia
...
         routeHandler.handle(route, request);
-        return;
+        handled = true;
+        break;
+      }
+    }
+    if (!handled) {
+      this._browserContext._onRoute(route, request);
+    } else  {
+      for (let i = this._routes.length - 1; i >= 0; i--) {
```
Comment (aslushnikov) [inline src/client/page.ts:183]:
```js
this._routes = this._routes.filter(route => !route.expired())
```

File: src/client/network.ts [hunk 2]
```diff
@@ -626,15 +626,18 @@ export class RouteHandler {
     this.handler = handler;
   }
 
+  public used(): boolean {
```
Comment (aslushnikov) [inline src/client/network.ts:629]: s/used/expired

File: src/client/browserContext.ts [hunk 3]
```diff
@@ -141,14 +141,23 @@ export class BrowserContext extends ChannelOwner<channels.BrowserContextChannel,
...
-        return;
+        handled = true;
+        break;
+      }
+    }
+    if (!handled) {
+      // it can race with BrowserContext.close() which then throws since its closed
+      route.continue().catch(() => {});
+    } else {
+      for (let i = this._routes.length - 1; i >= 0; i--) {
```
Comment (aslushnikov) [inline src/client/browserContext.ts:156]: filter here as well

**Review by aslushnikov — APPROVED** [review]

File: src/client/browserContext.ts [hunk 4]
```diff
@@ -141,14 +141,20 @@ export class BrowserContext extends ChannelOwner<channels.BrowserContextChannel,
...
+        break;
       }
     }
-    // it can race with BrowserContext.close() which then throws since its closed
-    route.continue().catch(() => {});
+    if (!handled) {
+      // it can race with BrowserContext.close() which then throws since its closed
+      route.continue().catch(() => {});
+    } else {
+      this._routes = this._routes.filter(route => !route.expired());
```
Comment (YusukeIwaki) [inline src/client/browserContext.ts:156]: @mxschmitt  Hi, I think `setNetworkInterceptionEnabled enabled: false)` can be called here when remaining handlers doesn't exist.

**Pull request merged**